util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -91,178 +114,179 @@ func NewWorker() *Worker {
 
 // Start starts the worker.
 func (w *Worker) Start() {
-	w.stopCh = make(chan struct{})
-	// TODO(#1631): make controller and worker communicate using k8s API (with CustomResources)
-	// instead of http.
-	w.listenToController()
+	w.populatePodName()
+	w.populateDynamicResource()
+	w.watchForWork()
 }
 
-func (w *Worker) listenToController() {
-	handlers := handlersMap{
-		"/startTCPServer":  w.StartTCPServer,
-		"/startTCPClient":  w.StartTCPClient,
-		"/startUDPServer":  w.StartUDPServer,
-		"/startUDPClient":  w.StartUDPClient,
-		"/startHTTPServer": w.StartHTTPServer,
-		"/startHTTPClient": w.StartHTTPClient,
-		"/metrics":         w.Metrics,
+func (w *Worker) populatePodName() {
+	var isPodNameAvailable bool
+	w.podName, isPodNameAvailable = os.LookupEnv("POD_NAME")
+	if !isPodNameAvailable {
+		w.updateStatus(map[string]interface{}{"error": "Pod name not set as environment variable"})
+		klog.Fatalf("Pod name not set as environment variable")
 	}
-	w.startListening(workerListenPort, handlers)
+	klog.Info("Pod Name set:", w.podName)
 }
 
-func (w *Worker) startListening(port string, handlers handlersMap) {
-	for urlPath, handler := range handlers {
-		http.HandleFunc(urlPath, handler)
-	}
-	if err := http.ListenAndServe(":"+port, nil); err != nil {
-		klog.Fatalf("Failed to start http server on port %v: %v", port, err)
+func (w *Worker) populateDynamicResource() {
+	gvk := schema.GroupVersionKind{Group: "clusterloader.io", Kind: "NetworkTestRequest", Version: "v1"}
+	var err error
+	w.dynamicResource, err = getDynamicResource(gvk, "netperf")
+	if err != nil {
+		w.updateStatus(map[string]interface{}{"error": fmt.Sprintf("Error getting Dynamic client for pod %s: %s", w.podName, err)})
+		klog.Fatal(err)
 	}
+	klog.Info("Dynamic resource created:", gvk)
 }
 
-// Metrics returns the metrics collected.
-func (w *Worker) Metrics(rw http.ResponseWriter, request *http.Request) {
-	select {
-	case <-w.stopCh:
-		if w.err != nil {
-			message := fmt.Sprintf("metrics collection failed: %v", w.err)
-			w.sendResponse(rw, http.StatusInternalServerError, Status{Message: message})
-			return
+func (w *Worker) watchForWork() {
+	watcher, err := w.getWatcher()
```

Comment:
That's not going to work universally:
1) watch starts at now, so if the object was created earlier you will never watch it
2) you don't handle all corner cases, resuming watches on timeouts/errors etc.

Instead of that - you should simply use the informer that will do all of that for you (and just register callbacks).
Here you have an example how to do that for dynamic clients:
https://github.com/kubernetes/perf-tests/blob/a0288cbf03d9bbf9ea5775182482f90dcf5f3d8a/clusterloader2/pkg/measurement/util/informer/informer.go#L54

Reply:
Modified to use informer.